EpikCord/client/http.py

```diff
@@ -0,0 +1,296 @@
+
+        url = self.clean_url(url)
+        bucket = self.buckets.get(f"{method}:{url}") or MockBucket()
+
+        await self.global_event.wait()
+        await bucket.wait()
+
+        for _ in range(5):
+            async with self.session.request(method, url, *args, **kwargs) as response:
+                if isinstance(bucket, MockBucket):
+                    bucket = await self.set_bucket(
+                        response=response,
+                        **route.major_parameters,
+                    )
+
+                data = await self.extract_content(response)
+
+                await self.log_request(response, data)
+
+                if response.status == 429:
+                    await self.handle_ratelimit(data, bucket)
+                elif response.headers.get("X-RateLimit-Remaining", "1") == "0":
+                    await self.handle_exhausted_bucket(response, bucket)
+                elif not response.ok:
+                    error = self.error_mapping.get(response.status, HTTPException)
+                    raise error(response, data)
+                elif response.ok:
+                    return response
+        raise TooManyRetries(f"Attempted {method} request to {url} 5 times but failed.")
+
+    async def set_bucket(
+        self,
+        *,
+        response: aiohttp.ClientResponse,
+        channel_id: Optional[int] = None,
+        guild_id: Optional[int] = None,
+        webhook_id: Optional[int] = None,
+        webhook_token: Optional[str] = None,
+    ) -> Union[Bucket, TopLevelBucket]:
+        url = response.url
+        method = response.method
+
+        if channel_id or guild_id or webhook_id or webhook_token:
+            bucket = TopLevelBucket(
+                channel_id=channel_id,
+                guild_id=guild_id,
+                webhook_id=webhook_id,
+                webhook_token=webhook_token,
+                )
+
+        else:
+
+            bucket = Bucket(bucket_hash=response.headers["X-RateLimit-Bucket"])
+            if bucket in self.buckets.values():
+                self.buckets[f"{method}:{url}"] = list(self.buckets.values())[
+                    list(self.buckets.values()).index(bucket)
+                ]
+                bucket = self.buckets[f"{method}:{url}"]
+            self.buckets[f"{method}:{url}"] = bucket
+
+        return bucket
+
+    async def extract_content(self, response: aiohttp.ClientResponse) -> Dict[str, Any]:
+        if response.headers["Content-Type"] == "application/json":
+            data = await response.json()
+        else:
+            data = {}
+        return data
+
+    async def handle_exhausted_bucket(
+        self, response: aiohttp.ClientResponse, bucket: Union[Bucket, TopLevelBucket]
+    ):
+        bucket.clear()
+        await asyncio.sleep(int(response.headers["X-RateLimit-Reset-After"]))
+        bucket.set()
+
+    async def handle_ratelimit(
+        self, data: Dict[str, Any], bucket: Union[Bucket, TopLevelBucket]
```

Comment:
Cant make the bucket argument something better than `Bucket` & `TopLevelBucket`? Like introducing a type that share the methods, a interface or something else?